### DISCUSSION TITLE
SLACK CLONE: How to use a cleanup function in a useEffect

### DISCUSSION THREAD

**Discussion by Gabrielle-Tobermann** [nss-evening-cohort-14/evening-client-side#236] [Help Me!] [answered]
Gabby Tobermann 

Other tickets? No. 

Issue: I am working on displaying messages for a specific DM. I've got a useEffect  with a promise that fetches the messages from Firebase. The response from the promise returns the correct data. But, I am getting an error that is preventing the messages to display. From what I understand, the component is mounting before the response from the promise is returned. 

error: Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function. 

I've looked into the cleanup function in a useEffect but I'm not sure how to implement it in this case. I'm not sure I quite understand this concept. 
https://maxrozen.com/demystifying-useeffect-cleanup-function
https://stackoverflow.com/questions/59780268/cleanup-memory-leaks-on-an-unmounted-component-in-react-hooks
https://reactjs.org/docs/hooks-effect.html

I've tried creating a mounted boolean (let mounted = true at the top of the useEffect) and then setting it to false in a return at the bottom of the useEffect. 

Link to code: 
component: https://github.com/BenJarrett/slack-clone/blob/gt-DM-messages/src/views/Conversation.js
data: https://github.com/BenJarrett/slack-clone/blob/gt-DM-messages/src/helpers/data/directMessageData.js

**Comment by drteresavasquez** [comment]
For your useParams, what is the value of that when logged?

**Reply by drteresavasquez** [reply]
You need to destructure. See this article: https://www.geeksforgeeks.org/reactjs-useparams-hook/

**Comment by Gabrielle-Tobermann** [comment, marked as answer]
Solution: I refactored our data so that the messages would have the ID of the channel or the DM. (So messages will have a key of communicationID which will either be a channel or DM firebaseKey). Then, I added a rule in firebase to filter messages by that ID. Then, I can just call my getMessages promise in the useEffect.
